code/game/objects/structures/training_machine.dm

```diff
@@ -0,0 +1,310 @@
+	desc = "Used for combat training simulations. Accepts standard training targets. A pair of buckling straps are attached."
+	icon = 'icons/obj/objects.dmi'
+	icon_state = "training_machine"
+	can_buckle = TRUE
+	buckle_lying = 0
+	max_integrity = 200
+	///Is the machine moving? Setting this to FALSE will automatically call stop_moving()
+	var/moving = FALSE
+	///The distance the machine is allowed to roam from its starting point
+	var/range = 1
+	///A random spot within range that the trainer is trying to reach
+	var/turf/target_position
+	///The turf the machine was on when it was activated
+	var/turf/starting_turf
+	///How fast the machine moves. Cannot be higher than MAX_SPEED
+	var/move_speed = 1
+	///Reference to a potentially attached object, either a target, trainer toolbox, or syndicate toolbox
+	var/obj/item/attached_item
+	///Time between attacks when emagged
+	var/attack_cooldown = 50
+	///Helper for timing attacks when emagged
+	var/last_attack_time = 0
+
+/obj/structure/training_machine/Destroy()
+	remove_attached_item(throwing = TRUE)
+	explosion(src, 0,0,1, flame_range = 2)
+	return ..()
+
+/obj/structure/training_machine/ui_state(mob/user)
+	return GLOB.physical_state
+
+/obj/structure/training_machine/ui_interact(mob/user, datum/tgui/ui)
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "TrainingMachine", name)
+		ui.open()
+
+/obj/structure/training_machine/ui_data(mob/user)
+	var/list/data = list()
+	data["range"] = range
+	data["movespeed"] = move_speed
+	data["moving"] = moving
+	return data
+
+/obj/structure/training_machine/ui_act(action, params)
+	if(..())
+		return
+	if (moving && obj_flags & EMAGGED)
+		visible_message("<span class='warning'>The [src]'s control panel fizzles slightly.</span>")
+		return
+	switch(action)
+		if("toggle")
+			toggle()
+			. = TRUE
+		if("range")
+			var/range_input = params["range"]
+			range = clamp(range_input, MIN_RANGE, MAX_RANGE)
+			. = TRUE
+		if("movespeed")
+			var/range_input = params["movespeed"]
+			move_speed = clamp(range_input, MIN_SPEED, MAX_SPEED)
+			. = TRUE
+
+/obj/structure/training_machine/attack_hand(mob/user)
+	ui_interact(user)
+
+/obj/structure/training_machine/attackby(obj/item/target, mob/user)
+	. = ..()
+	if (user.a_intent == INTENT_HARM)
+		return
+	if (!istype(target, /obj/item/training_toolbox) && !istype(target, /obj/item/target))
+		return
+	if (length(buckled_mobs))
+		return
+	if (obj_flags & EMAGGED)
+		to_chat(user, "<span class='warning'>The toolbox is somehow stuck on! It won't budge!</span>")
+		return
+	attach_item(target)
+	to_chat(user, "<span class='notice'>You attach \the [attached_item] to the training device.</span>")
+	playsound(src, "rustle", 50, TRUE)
+
+/obj/structure/training_machine/proc/attach_item(target)
+	remove_attached_item()
+	attached_item = target
+	attached_item.forceMove(src)
+	attached_item.vis_flags |= VIS_INHERIT_ID
+	vis_contents += attached_item
```

Comment:
Could you register to the `attached_item`'s `COMSIG_PARENT_QDELETING` event and handle it properly?